### PR TITLE
cudatoolkit: set cmake flag CUDA_HOST_COMPILER to gcc

### LINKED ISSUE #109283: ethminer: CUDA code fails to build
**Describe the bug**
I'm not sure what's exactly going on here, but it looks like the build is attempting to use gcc10 includes from somewhere, while the default nvcc from cudatoolkit only supports gcc9 at best. Maybe someone better versed with how cmake works on NixOS could take a look?

```
...
[ 57%] Building NVCC (Device) object libethash-cuda/CMakeFiles/ethash-cuda.dir/ethash-cuda_generated_ethash_cuda_miner_kernel.cu.o
4 warnings generated.
/nix/store/5wfh94xwg2hd7n82j8klgn4nmf07is24-gcc-10.2.0/include/c++/10.2.0/type_traits(1396): error: type name is not allowed
...
```

**To Reproduce**
1. Try to build `pkgs.ethminer` from nixpkgs (213286bb181c9b6588bb5c4fa130301cefc90dd0).

**Expected behavior**
The package builds successfully with CUDA support.

**Additional context**
Persists after version bump

**Notify maintainers**

cc @nand0p @breakds 

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: pkgs.ethminer
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by fx-chun**
###### Motivation for this change
It seems that the FindCUDA module in cmake was populating nvcc's `-ccbin` with the latest gcc, causing cmake projects using the module to try and use the newer gcc (rather than the version specified in cudatoolkit) with nvcc to build code. 

It looks like this behavior was OK with gcc9 since that's what we used anyways, but with the gcc10 merge I imagine this has broken builds depending on this module (it has at least broken ethminer). This PR explicitly sets the relevant cmake option to force nvcc to use a compatible gcc.

This PR also fixes ethminer by not propagating clang flags to nvcc and bumps ethminer to v0.19.0.

Closes #109283.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by noneucat** [comment]
cc @nand0p

**Review by r-burns — COMMENTED** [review]
I don't know why this package is using clangStdenv but I don't think it's needed anymore - with the CUDA_HOST_COMPILER flag this seems too build with the default gcc 10 stdenv

**Comment by wheelsandmetal** [comment]
> I don't know why this package is using clangStdenv

The reason for the change is https://github.com/NixOS/nixpkgs/pull/107239

Not sure if that can be reverted when this is merged?

**Comment by SuperSandro2000** [comment]
ping @danieldk

File: pkgs/development/compilers/cudatoolkit/common.nix [hunk 1]
```diff
@@ -147,6 +147,9 @@ stdenv.mkDerivation rec {
     mkdir -p $out/nix-support
     echo "cmakeFlags+=' -DCUDA_TOOLKIT_ROOT_DIR=$out'" >> $out/nix-support/setup-hook
 
+    # Force cmake to build with our gcc version for nvcc
```
Comment (danieldk) [inline pkgs/development/compilers/cudatoolkit/common.nix:150]:
```suggestion
    # Set the host compiler to be used by nvcc for CMake-based projects:
    # https://cmake.org/cmake/help/latest/module/FindCUDA.html#input-variables
```

**Review by danieldk — CHANGES_REQUESTED** [review]
Good catch! So, the wrapping of `nvcc` does not work with CMake because `FindCUDA` passes `-ccbin` to override the compiler.

I added a small requested change, so that it is a bit clearer what the flag does and where it comes from.

**Comment by SuperSandro2000** [comment]
> Is there anything I can do to help this get merged? It's a little out my league, but ethminer is still broken and I would love to use it.

You can fix the merge conflict.

**Comment by noneucat** [comment]
Resolved, thank you!
@SuperSandro2000 @danieldk

**Review by Atemu — COMMENTED** [review]
I can't really comment on CUDA stuff because I sold all of my Nvidia GPUs but don't you need to set the stdenv back to the regular non-clang one too?

File: pkgs/tools/misc/ethminer/default.nix [hunk 2]
```diff
@@ -45,6 +45,7 @@ in stdenv.mkDerivation rec {
     "-DCMAKE_BUILD_TYPE=Release"
   ] ++ lib.optionals (!cudaSupport) [
     "-DETHASHCUDA=OFF" # on by default
+    "-DCUDA_PROPAGATE_HOST_FLAGS=off"
```
Comment (albertov) [inline pkgs/tools/misc/ethminer/default.nix:48]: This line needs to go after line 45 so it is passed when cuda is enabled, not when it isn't. Also, remove line 85, the `broken = cudaSupport` one.

**Review by albertov — CHANGES_REQUESTED** [review]
This patch fixed ethminer in my system but I had to make some tweaks to it. See comments

File: pkgs/tools/misc/ethminer/default.nix (hunk 2 above)
Comment (Atemu) [reply pkgs/tools/misc/ethminer/default.nix:48]: Would probably be best to just put all of that in a ternary; i.e.

`++ if cudaSupport then [ ... ] else [ ... ];

Comment (noneucat) [reply pkgs/tools/misc/ethminer/default.nix:48]: Overlooked, sorry :)

**Review by Atemu — APPROVED** [review]
Diff looks good but I can't test CUDA stuff.

**Comment by danieldk** [comment]
> Diff looks good but I can't test CUDA stuff.

I have an NVIDIA GPU, but I don't want to run anything mining-related, so I can't test. Looks good to me otherwise.

**Comment by Atemu** [comment]
The built-in benchmark is completely offline and running it for a few seconds won't hurt your GPU in any way fyi.

**Comment by trepetti** [comment]
I was able to test this and ran the benchmark without issue. Wasn't sure what to provide for the integer, since it is pretty far outside my wheelhouse, but it compiled (the main point of the change set) and ran fine.

**Pull request merged**